### ISSUE TITLE
What events are streamed from cache.download()?

### ISSUE THREAD

**Issue by chmac** [markmarijnissen/cordova-file-cache#20]
Firstly, thanks for an awesome plugin. It came highly recommended and seems to be precisely what we're looking for.

I'm wondering if there are docs somewhere about what type of events the `onprogress` callback gets passed?

Specifically, I'm thinking about how to implement a "ready" event for each file. Ideally I'd like to throw some files at the cache, and then trigger a callback when those files have been successfully fetched. My first thought was to try listening to the progress events and then reacting appropriately. Does that approach make sense?

**Comment by markmarijnissen** [comment]
Yes it does! Should be possible with existing api, although you must
correlate download percent/index with index in download queue yourself.
Could be a pull request to add a list of downloaded files to the events

Op di 30 aug. 2016 23:54 schreef Callum Macdonald <notifications@github.com

**Comment by chmac** [comment]
@markmarijnissen I've been playing around with the `onprogress` callback, trying different combinations of several files, single files, and `true` or `false` for `includeFileProgress`. I can't quite figure out what all the various properties of the events mean, and how I should interpret them.

Are there docs on this somewhere? If not, would you be up for writing a little about what each property means, and what events are streamed?

Here's an example event I'm seeing:

```
ProgressEvent
  bubbles: false
  cancelBubble: false
  cancelable: false
  lengthComputable: false
  loaded: 0
  path: "cache/bar.json"
  percentage: 0.25
  queueIndex: 0
  queueSize: 4
  target: null
  total: 0
  type: undefined
  url: "https://s3.eu-central-1.amazonaws.com/foo/bar.json"
  ProgressEvent Prototype
```

At a guess, it looks like `percentage` is how far through the total queue we currently are. So with a queueSize of 4 and 1 of those 4 downloads completed, then `percentage = 0.25`. Is that correct?

Am I correct in thinking that one event will be streamed for the successful download of each file? Are the `url` and `path` properties also unique to that file? I'm seeing the last file in my queue generating 2 events, but that's fairly easy to handle.

Bottom line, I'm struggling to grasp how I can put a single file into the cache, and then trigger a callback once that file has been successfully loaded into the cache. I looked into the code and am slightly nervous that the `.download()` method appears to call `.abort()` when it starts, so calling it multiple times could potentially kill downloads which are currently in progress. I'm therefore guessing I need some kind of mutual exclusion in my own code where I ensure that I only ever have 1 call to `.download()` at a time.
